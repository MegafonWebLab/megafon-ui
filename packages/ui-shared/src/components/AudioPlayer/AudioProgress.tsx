import React from 'react';
import { cnCreate } from '@megafon/ui-helpers';
import AudioRange from './AudioRange';
import { timerFormat } from './utils';
import './style/AudioProgress.less';

export const INTERVAL_DELAY = 1000;

export interface IAudioProgressProps {
    audioRef: React.MutableRefObject<HTMLAudioElement | null>;
    audioTitle: string;
    isPlaying: boolean;
    isPause: boolean;
    onChangeAudioCurrentTime: (currentTime: number) => void;
    onPLay: () => void;
    onSetIsPlaying: (value: boolean) => void;
}

const cn = cnCreate('mfui-audio-progress');
const AudioProgress: React.FC<IAudioProgressProps> = ({
    audioRef,
    audioTitle,
    isPlaying,
    isPause,
    onChangeAudioCurrentTime,
    onPLay,
    onSetIsPlaying,
}) => {
    const intervalId = React.useRef<NodeJS.Timeout | null>(null);

    const [trackDuration, setTrackDuration] = React.useState<number>(0);
    const [trackProgress, setTrackProgress] = React.useState<number>(0);

    const progressColorPercent: number = React.useMemo(
        () => (trackProgress / trackDuration) * 100,
        [trackDuration, trackProgress],
    );
    const progressValue: number = React.useMemo(
        () => (isPlaying || isPause ? trackProgress : trackDuration),
        [isPause, isPlaying, trackDuration, trackProgress],
    );

    const handleStartTimer = React.useCallback(() => {
        intervalId.current && clearInterval(intervalId.current);

        intervalId.current = setInterval(() => {
            if (audioRef.current?.ended) {
                onSetIsPlaying(false);
                setTrackProgress(Math.floor(audioRef.current?.duration || 0));

                return;
            }

            setTrackProgress(Math.floor(audioRef.current?.currentTime || 0));
        }, INTERVAL_DELAY);
    }, [audioRef, onSetIsPlaying]);

    const handleScrubProgress = (e: React.ChangeEvent<HTMLInputElement>) => {
        const currentTime = Number(e.target.value);

        setTrackProgress(currentTime);
        onChangeAudioCurrentTime(currentTime);
    };

    const handleScrubEndProgress = () => {
        if (!isPlaying) {
            onPLay();
        }

        handleStartTimer();
    };

    React.useEffect(() => {
        setTrackProgress(Math.floor(audioRef?.current?.currentTime || 0));
    }, [audioRef]);

    React.useEffect(() => {
        const audioNode = audioRef.current;

        if (audioNode) {
            audioNode.onloadedmetadata = () => {
                setTrackDuration(Math.round(audioNode.duration));
            };
        }
    }, [audioRef]);

    React.useEffect(() => {
        if (isPlaying) {
            handleStartTimer();
        }

        if (isPause) {
            intervalId.current && clearInterval(intervalId.current);
        }
    }, [handleStartTimer, isPause, isPlaying]);

    React.useEffect(
        () => (): void => {
            intervalId.current && clearInterval(intervalId.current);
        },
        [],
    );

    return (
        <div className={cn()}>
            <div className={cn('info')}>
                <span className={cn('title')}>{audioTitle}</span>
                <span className={cn('value')}>{timerFormat(progressValue)}</span>
            </div>
            <div className={cn('range-wrapper')}>
                <AudioRange
                    className={cn('range')}
                    maxValue={trackDuration}
                    value={trackProgress}
                    step="1"
                    colorPercent={progressColorPercent || 0}
                    onChange={handleScrubProgress}
                    onMouseUp={handleScrubEndProgress}
                    onTouchEnd={handleScrubEndProgress}
                />
            </div>
        </div>
    );
};

export default AudioProgress;
