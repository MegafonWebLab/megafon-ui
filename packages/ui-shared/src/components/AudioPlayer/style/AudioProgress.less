@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-audio-progress');

@{b} {
    width: 100%;
    margin-right: 16px;

    &__info {
        display: flex;
        align-items: center;
        justify-content: space-between;
    }

    &__value {
        color: var(-spbSky3);

        .smallFont();
    }

    &__range-wrapper {
        margin-top: -2px;
    }

    &__title {
        color: var(--stcBlack);

        .smallFont();
    }
}
