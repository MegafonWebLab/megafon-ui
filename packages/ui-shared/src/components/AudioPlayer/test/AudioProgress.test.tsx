import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import AudioProgress, { IAudioProgressProps } from '../AudioProgress';
import type { IAudioRangeProps } from '../AudioRange';

jest.useFakeTimers();

const audioRefMock = {
    current: {
        currentTime: 10,
        duration: 100,
        ended: false,
    },
} as React.MutableRefObject<HTMLAudioElement>;

const props: IAudioProgressProps = {
    audioRef: audioRefMock,
    audioTitle: 'test audioTitle',
    isPlaying: false,
    isPause: false,
    onChangeAudioCurrentTime: jest.fn(),
    onPLay: jest.fn(),
    onSetIsPlaying: jest.fn(),
};

const setTrackDuration = jest.fn();
const setTrackProgress = jest.fn();

const setMockedState = ({ trackDuration = 0, trackProgress = 0 } = {}): void => {
    jest.spyOn(React, 'useState').mockImplementationOnce(() => [trackDuration, setTrackDuration]);
    jest.spyOn(React, 'useState').mockImplementationOnce(() => [trackProgress, setTrackProgress]);
};

const getWrapper = (additionalProps?: Partial<IAudioProgressProps>): ShallowWrapper =>
    shallow(<AudioProgress {...props} {...additionalProps} />);

describe('<AudioProgress />', () => {
    const mockUseEffect = (): void => {
        jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
    };

    const mockUseRef = (intervalId: number): void => {
        jest.spyOn(React, 'useRef').mockImplementationOnce(() => ({ current: intervalId }));
    };

    afterEach(() => jest.clearAllMocks());
    afterAll(() => jest.restoreAllMocks());

    describe('snapshots', () => {
        it('should render component', () => {
            setMockedState();

            const wrapper = getWrapper();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handleScrubEndProgress', () => {
        it('should call handleStartTimer and call onPLay when isPlaying to false', () => {
            setMockedState();

            const wrapper = getWrapper();
            jest.runOnlyPendingTimers();
            const AudioRangeProps = wrapper.find('AudioRange').props() as IAudioRangeProps;
            AudioRangeProps.onMouseUp && AudioRangeProps.onMouseUp();

            expect(setInterval).toBeCalledWith(expect.any(Function), 1000);
            expect(props.onPLay).toBeCalled();
        });

        it('should call clearInterval and not call handlePLay', () => {
            const intervalId = 1234;

            mockUseRef(intervalId);
            setMockedState();

            const wrapper = getWrapper({ isPlaying: true });
            const AudioRangeProps = wrapper.find('AudioRange').props() as IAudioRangeProps;
            AudioRangeProps.onMouseUp && AudioRangeProps.onMouseUp();

            expect(clearInterval).toBeCalledWith(intervalId);
            expect(props.onPLay).not.toBeCalled();
        });
    });

    describe('useEffect', () => {
        beforeEach(() => {
            mockUseEffect();
            mockUseEffect();
            mockUseEffect();
        });

        it('should call handleStartTimer when isPlaying to true', () => {
            setMockedState();

            getWrapper({ isPlaying: true });
            jest.runOnlyPendingTimers();

            expect(setInterval).toBeCalledWith(expect.any(Function), 1000);
        });

        it('should call clearInterval when isPause to true', () => {
            const intervalId = 1234;

            mockUseRef(intervalId);
            setMockedState();

            getWrapper({ isPause: true });

            expect(clearInterval).toBeCalledWith(intervalId);
        });
    });

    describe('useEffectCallback', () => {
        let cleanupFunc: ReturnType<React.EffectCallback>;

        beforeEach(() => {
            jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
            jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
            jest.spyOn(React, 'useEffect').mockImplementation(func => {
                cleanupFunc = func();
            });
        });

        it('should call clearInterval', () => {
            const intervalId = 1234;

            mockUseRef(intervalId);
            setMockedState();

            const wrapper = getWrapper();

            cleanupFunc && cleanupFunc();
            wrapper.unmount();

            expect(clearInterval).toBeCalledWith(intervalId);
        });
    });
});
