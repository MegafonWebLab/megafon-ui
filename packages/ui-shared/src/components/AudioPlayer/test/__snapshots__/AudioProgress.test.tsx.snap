// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioProgress /> should render component with default state 1`] = `
<div>
  <div
    class="mfui-audio-progress"
  >
    <div
      class="mfui-audio-progress__info"
    >
      <span
        class="mfui-audio-progress__title"
      >
        test audioTitle
      </span>
      <span
        class="mfui-audio-progress__value"
      >
        01:40
      </span>
    </div>
    <div
      class="mfui-audio-progress__range-wrapper"
    >
      <input
        class="mfui-audio-range mfui-audio-progress__range"
        data-testid="AudioTimeRange"
        max="100"
        min="0"
        step="1"
        style="background-size: 0% 100%;"
        type="range"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`<AudioProgress /> should render with playing state 1`] = `
<div>
  <div
    class="mfui-audio-progress"
  >
    <div
      class="mfui-audio-progress__info"
    >
      <span
        class="mfui-audio-progress__title"
      >
        test audioTitle
      </span>
      <span
        class="mfui-audio-progress__value"
      >
        00:10
      </span>
    </div>
    <div
      class="mfui-audio-progress__range-wrapper"
    >
      <input
        class="mfui-audio-range mfui-audio-progress__range"
        data-testid="AudioTimeRange"
        max="100"
        min="0"
        step="1"
        style="background-size: 10% 100%;"
        type="range"
        value="10"
      />
    </div>
  </div>
</div>
`;
